crates/task-impls/src/consensus/helpers.rs

```diff
@@ -539,6 +546,54 @@ pub async fn publish_proposal_if_able<TYPES: NodeType>(
     .await
 }
 
+/// Trigger a request to the network for a proposal for a view and wait for the response
+#[cfg(not(feature = "dependency-tasks"))]
+async fn fetch_proposal<TYPES: NodeType>(
+    view: TYPES::Time,
+    event_stream: Sender<Arc<HotShotEvent<TYPES>>>,
+    quorum_membership: Arc<TYPES::Membership>,
+    consensus: Arc<RwLock<Consensus<TYPES>>>,
+) -> Result<Leaf<TYPES>> {
+    let (tx, mut rx) = broadcast(1);
+    let event = ProposalMissing {
+        view,
+        response_chan: tx,
+    };
+    broadcast_event(
+        Arc::new(HotShotEvent::QuorumProposalMissing(event)),
+        &event_stream,
+    )
+    .await;
+    let Ok(Ok(Some(proposal))) = async_timeout(REQUEST_TIMEOUT, rx.recv_direct()).await else {
+        bail!("Request for proposal failed");
+    };
+    let view = proposal.data.view_number();
+    let justify_qc = proposal.data.justify_qc.clone();
+
+    if !justify_qc.is_valid_cert(quorum_membership.as_ref()) {
+        bail!("Invalid justify_qc in proposal for view {}", *view);
+    }
+    let mut consensus_write = consensus.write().await;
+    let leaf = Leaf::from_quorum_proposal(&proposal.data);
+    let state = Arc::new(
+        <TYPES::ValidatedState as ValidatedState<TYPES>>::from_header(&proposal.data.block_header),
+    );
+
+    consensus_write.update_validated_state_map(
+        view,
+        View {
+            view_inner: ViewInner::Leaf {
+                leaf: leaf.commit(),
+                state,
+                delta: None,
+            },
+        },
+    );
```

Comment:
I'm not really able to review this part, because I don't know exactly what checks we need to make before calling this function.

In e.g. `update_state_and_vote_if_able` we seem to gate calling this on validating the DAC, so I don't know if that's a requirement or if there are other things we need to be checking for safety.